flink-table/flink-table-planner/src/main/java/org/apache/calcite/sql/fun/SqlCastFunction.java

```diff
@@ -98,7 +103,21 @@ public RelDataType inferReturnType(SqlOperatorBinding opBinding) {
         assert opBinding.getOperandCount() == 2;
         RelDataType ret = opBinding.getOperandType(1);
         RelDataType firstType = opBinding.getOperandType(0);
-        ret = opBinding.getTypeFactory().createTypeWithNullability(ret, firstType.isNullable());
+
+        RelOptCluster relOptCluster =
+                ((FlinkCalciteSqlValidator) ((SqlCallBinding) opBinding).getValidator())
+                        .getRelOptCluster();
+        TableConfig tableConfig = ShortcutUtils.unwrapContext(relOptCluster).getTableConfig();
+        boolean legacyCastEnabled =
+                tableConfig
+                        .get(ExecutionConfigOptions.TABLE_EXEC_LEGACY_CAST_BEHAVIOUR)
+                        .isEnabled();
+        ret =
+                opBinding
+                        .getTypeFactory()
+                        .createTypeWithNullability(
+                                ret, firstType.isNullable() || legacyCastEnabled);
```

Comment:
Please include a note about this behavior in the class comment as well.

Reply:
added